_includes/tutorials/kafka-producer-application-callback/kafka/markup/dev/make-application.adoc

```diff
@@ -0,0 +1,86 @@
+
+(In practice you may want to use a dependency injection framework library, such as the  https://spring.io/projects/spring-framework[Spring Framework]).
+
+
+Next let's take a look at the `KafkaProducerCallbackApplication.produce` method
+[source, java]
+.KafkaProducerCallbackApplication.produce
+----
+public void produce(final String message) {
+    final String[] parts = message.split("-");  <1>
+    final String key, value;
+    if (parts.length > 1) {
+      key = parts[0];
+      value = parts[1];
+    } else {
+      key = "NO-KEY";
+      value = parts[0];
+    }
+    final ProducerRecord<String, String> producerRecord = new ProducerRecord<>(outTopic, key, value);  <2>
+    producer.send(producerRecord, (recordMetadata, exception) -> {  <3>
+              if (exception == null) {                          <4>
+                  System.out.println("Record written to offset " +
+                          recordMetadata.offset() + " timestamp " +
+                          recordMetadata.timestamp());
+              } else {
+                  System.err.println("An error occurred"); <5>
+                  exception.printStackTrace(System.err);
+              }
+        });
+  }
+
+----
+
+<1> Process the String for sending message
+<2> Create the `ProducerRecord`
+<3> Send the record to the broker specifying a `Callback` instance as a lambda function
+<4> If there's no exceptions print the offset and timestamp of the acknowleged record
+<5> Error handling portion-in this case printing the stacktrace to `System.err`
+
+The `KafkaProducerCallbackApplication.produce` method does some processing on a `String`, and then sends the https://kafka.apache.org/25/javadoc/org/apache/kafka/clients/producer/ProducerRecord.html[`ProducerRecord`].  While this code is a trivial example, it's enough to show the example of using a `KafkaProducer`.
+
+Notice that this overload of the https://javadoc.io/static/org.apache.kafka/kafka-clients/2.6.0/org/apache/kafka/clients/producer/KafkaProducer.html#send-org.apache.kafka.clients.producer.ProducerRecord-org.apache.kafka.clients.producer.Callback--[`KafkaProducer.send`] method accepts a second parameter, an instance of the https://kafka.apache.org/26/javadoc/org/apache/kafka/clients/producer/Callback.html[Callback] interface.
+
+The `Callback` provides a way of handling any actions you want to take on request completion *_asynchronously_*.  Note that the `Callback` code executes on the producer's I/O thread so any actions here should execute quickly.  Any time consuming tasks could cause a delay in sending new records.
+
+The `KafkaProducer.send` method is asynchronous and returns as soon as the provided record is placed in the buffer of records to be sent to the broker. Once the broker acknowledges that the record has been appended to its log, the broker completes the produce request, which the application receives as `RecordMetadata`—information about the committed message.
```

Comment:
Is the send method itself technically `asynchronous`?  Or the callback is asynchronous?  Not sure if this is a distinction worth making, but FYC maybe tweak as:

```
The `KafkaProducer.send` method returns as soon as the provided record is placed in the buffer of records to be sent to the broker. Once the broker acknowledges that the record has been appended to its log, the broker completes the produce request by asynchronously sending back to the producer a `RecordMetadata` object—information about the committed message.
```

(or something like this)

Reply:
The send method itself is asynchronous. The callback is executed on the same thread that executes the request. What you have could work as well, but I think I'm going to leave this as is.